### ISSUE TITLE
Laravel FFMPEG Add HLS Video Watermark issue

### ISSUE THREAD

**Issue by jefrisingh** [pascalbaljetmedia/laravel-ffmpeg#203]
I'm using "pbmedia/laravel-ffmpeg": "^7.0" in "laravel/framework": "^7.0" , I want to add water mark to the HLS video, Here is the example given in the github repo https://github.com/pascalbaljetmedia/laravel-ffmpeg

```
FFMpeg::open(['video.mp4', 'video2.mp4'])
    ->export()
    ->addFilter(function(ComplexFilters $filters) {
        // $filters->watermark(...);
    });
```
But It's not have enough information, Here is the working example everyware in the internet

```
addFilter(function ($filters) {
            $filters->watermark($watermarkPath, [
                'position' => 'relative',
                'bottom' => 50,
                'right' => 50,
            ]);
        })
```
But the watermark method has 2 more required parameters in the library

```
addFilter(function (ComplexFilters $filters) {
            $filters->watermark($in,$watermarkPath,$out, [
                'position' => 'relative',
                'bottom' => 50,
                'right' => 50,
            ]);
        })
```
**What should I pass for $in and $out** Here is the full code which I have

```
use FFMpeg;
use App\Video;
use FFMpeg\Format\Video\X264;
use FFMpeg\Filters\AdvancedMedia\ComplexFilters;

$low = (new X264('aac'))->setKiloBitrate(100);
$mid = (new X264('aac'))->setKiloBitrate(250);
$high = (new X264('aac'))->setKiloBitrate(500);


       FFMpeg::fromDisk('local')
        ->open($this->video->path)
        ->exportForHLS()
        ->addFilter(function (ComplexFilters $filters) {
            $filters->watermark(null,public_path('watermark.png'),null, [
                'position' => 'relative',
                'bottom' => 50,
                'right' => 50,
            ]);
        })
        ->addFormat($low)
        ->addFormat($mid)
        ->addFormat($high)
        ->save("public/videos/{$this->video->id}/{$this->video->id}.m3u8");
```
By passing null I can't able to see any watermark , If I remove those nulls , I'll have error
![image](https://user-images.githubusercontent.com/13378709/84088758-e36a2280-aa0a-11ea-9103-356d444ab187.png)
Please look at the required parameters in watermark function under ComplexFilters, Thanks :).

**Comment by mateuszBedzinski** [comment]
Same here.
Did you manage this? previously wersion works fine?

**Comment by jefrisingh** [comment]
Yes I did tested 2 years old laravel source code ("pbmedia/laravel-ffmpeg": "^4.0"), That is working without any issue, If I updated it to latest version I'm having this issue ,

**Comment by mateuszBedzinski** [comment]
I've switched to original **PHP-FFMpeg/PHP-FFMpeg** and there works fine (drawtex filter but you can do watermark same way) 

```
$moviePath  = Storage::disk('local')->path( $moviePath );
			
$text       = $titlesOrg[ $items[ 1 ] ];
$output     = Storage::disk('local')->path( $outputPath . '/' . $fileName );
	
$ffmpeg = FFMpeg::create();

$video = $ffmpeg->open( $moviePath );

$command = "text='$text': fontfile=OpenSans-Regular.ttf: fontcolor=white: fontsize=120: box=0: boxcolor=black@0.5: 
 boxborderw=5: x=50: y=100:";

$video->filters()->custom( "drawtext=$command" );

$video->save( $format, $output );
```

**Comment by pascalbaljet** [comment]
To use a watermark filter in a HLS export with v7.0, you need to upgrade your code. This is mentioned in the [upgrade section](https://github.com/pascalbaljetmedia/laravel-ffmpeg#upgrading-to-v7).

```php
FFMpeg::open('video.mp4')
    ->exportForHLS()

    // option A (highly recommended)
    ->addFormat($format, function ($media) {
        $media->addFilter(function ($filters, $in, $out) {
            $imagePath = '';
            $coordinates = [];

            $filters->watermark($in, $imagePath, $out, $coordinates);
        });
    })

    // option B
    ->addFormat($format, function($media) {
        $media->addLegacyFilter(function ($filters) {
            $imagePath = '';
            $coordinates = [];

            $filters->watermark($imagePath, $coordinates);
        });
    })

    ->save('adaptive.m3u8');
```

**Comment by mohsen-zanjani** [comment]
Hello
I want to have two filters when exporting for HLS (drawtext and resize) ...
But all the following cases gave an error (error is mentioned in the comment).
This is while each one works properly on its own.

```php
// create video formats ...
$lowBitRate = (new X264('aac', 'libx264'))->setKiloBitrate(400);      // 426x240 pixels
$midBitRate = (new X264('aac', 'libx264'))->setKiloBitrate(800);      // 640x360 pixels
$highBitRate = (new X264('aac', 'libx264'))->setKiloBitrate(1200);     // 854x480 pixels
$superBitRate = (new X264('aac', 'libx264'))->setKiloBitrate(1500);     // 1280x720 pixels

FFMpeg::fromDisk('public')
    ->open('IMG_5316.MP4')
    ->exportForHLS()
    ->addFormat($lowBitRate, function ($media) {
        // Error: Call to undefined method FFMpeg\Filters\AdvancedMedia\ComplexFilters::resize()
        $media->addFilter(function ($filters) {
            $filters->resize(new Dimension(426, 240));
            $filters->custom("drawtext=text='Watermark Text':x=10:y=h-th-10:fontfile=tahoma.ttf:fontsize=16:fontcolor=white"); // $in, $parameters, $out
        });

        // Error: Call to undefined method FFMpeg\Filters\AdvancedMedia\ComplexFilters::resize()
        $media->addFilter("drawtext=text='Watermark Text':x=10:y=h-th-10:fontfile=tahoma.ttf:fontsize=16:fontcolor=white"); // $in, $parameters, $out
        $media->addFilter(function ($filters) {
            $filters->resize(new Dimension(426, 240));
        });

        // Encoding failed
        $media->addFilter("drawtext=text='Watermark Text':x=10:y=h-th-10:fontfile=C:\\laragon\\www\\ffmpeg-new\\public\\fonts\\tahoma.ttf:fontsize=16:fontcolor=white"); // $in, $parameters, $out
        $media->addLegacyFilter(function ($filters) {
            $filters->resize(new Dimension(426, 240));
        });

        // Error: Call to undefined function scale=426:240()
         $media->addFilter('scale=426:240', "drawtext=text='Watermark Text':x=10:y=h-th-10:fontfile=tahoma.ttf:fontsize=16:fontcolor=white");

        // Error: FFMpeg\Exception\RuntimeException Encoding failed
         $media->addFilter('scale=426:240');
         $media->addFilter("drawtext=text='Watermark Text':x=10:y=h-th-10:fontfile=tahoma.ttf:fontsize=16:fontcolor=white");

        // Error: Array to string conversion
         $media->addFilter(['scale=426:240', "drawtext=text='Watermark Text':x=10:y=h-th-10:fontfile=tahoma.ttf:fontsize=16:fontcolor=white"]);
    })
    ->addFormat($midBitRate, function ($media) {
        $media->scale(640, 360);
    })
    ->addFormat($highBitRate, function ($media) {
        $media->addFilter(function ($filters, $in, $out) {
             $filters->custom($in, 'scale=854:480', $out); // $in, $parameters, $out
            // $filters->custom($in, "drawtext=text='Watermark Text':x=10:y=h-th-10:fontfile=tahoma.ttf:fontsize=16:fontcolor=white", $out); // $in, $parameters, $out
        });
    })
    ->addFormat($superBitRate, function ($media) {
        $media->addLegacyFilter(function ($filters) {
             $filters->resize(new \FFMpeg\Coordinate\Dimension(1280, 720));
            // $filters->custom(new CustomFilter("drawtext=text='Watermark Text':x=10:y=h-th-10:fontfile=tahoma.ttf:fontsize=16:fontcolor=white"));
        });
    })
    ->save('adaptive_steve.m3u8');

```

@pascalbaljet Please give me an example of how to do this

**Comment by pascalbaljet** [comment]
@mohsen-zanjani this should be fixed in [v7.4](https://github.com/protonemedia/laravel-ffmpeg/releases/tag/7.4.0) (see notes).

**Comment by mohsen-zanjani** [comment]
> @mohsen-zanjani this should be fixed in [v7.4](https://github.com/protonemedia/laravel-ffmpeg/releases/tag/7.4.0) (see notes).

Thanks for releasing the new version
I tested the `"HLS export with multiple filters per format"` feature and it worked fine. I also have a special thanks for `dd`.

Just one case: 
I got an error for some videos. (My special case is 360VR video)
 ` "Too many packets buffered for output stream 0: 1."`

I came up with the solution `"-max_muxing_queue_size 4000"` according to my searches.
Since I could not add this filter to the HLS output, I was forced to change the `addHLSParametersToFormat` method of the `HLSExporter` class as follows, and the error was fixed:
```php
private function addHLSParametersToFormat(DefaultVideo $format, string $segmentsPattern, Disk $disk)
{
	$format->setAdditionalParameters([
		'-sc_threshold',
		'0',
		'-g',
		$this->keyFrameInterval,
		'-hls_playlist_type',
		'vod',
		'-hls_time',
		$this->segmentLength,
		'-hls_segment_filename',
		$disk->makeMedia($segmentsPattern)->getLocalPath(),
		'-max_muxing_queue_size', // new command
		9999,
	]);
}
```
Of course, changing the package code is not the right thing to do, but my question is how can I add this command correctly?
